scio-smb/src/main/java/org/apache/beam/sdk/extensions/smb/SortedBucketTransform.java

```diff
@@ -0,0 +1,273 @@
+    this.sources = sources;
+    this.transformFn = transformFn;
+    this.bucketMetadata = bucketMetadata;
+  }
+
+  @Override
+  public final WriteResult expand(PBegin begin) {
+    final SourceSpec<FinalKeyT> sourceSpec = SortedBucketSource.getSourceSpec(finalKeyClass, sources);
+
+    Preconditions.checkArgument(
+        bucketMetadata.getNumBuckets() == sourceSpec.leastNumBuckets,
+        "Specified number of buckets %s does not match smallest bucket size among"
+            + " inputs: %s.", bucketMetadata.getNumBuckets(), sourceSpec.leastNumBuckets
+    );
+
+    final FileAssignment tempFileAssignment = filenamePolicy.forTempFiles(tempDirectory);
+
+    final Create.Values<Integer> createBuckets = Create.of(
+        IntStream.range(0, sourceSpec.leastNumBuckets).boxed().collect(Collectors.toList())
+    ).withCoder(VarIntCoder.of());
+
+    final Create.Values<ResourceId> writeTempMetadata =
+        SortedBucketSink.WriteTempFiles.writeMetadataTransform(tempFileAssignment, bucketMetadata);
+
+    @SuppressWarnings("deprecation")
+    final Reshuffle.ViaRandomKey<Integer> reshuffle = Reshuffle.viaRandomKey();
+
+    return PCollectionTuple
+        .of(
+            new TupleTag<>("TempMetadata"),
+            begin
+                .getPipeline()
+                .apply("WriteTempMetadata", writeTempMetadata)
+                .setCoder(ResourceIdCoder.of())
+        ).and(
+            new TupleTag<>("TempBuckets"),
+            begin.getPipeline()
+                .apply("CreateBuckets", createBuckets)
+                .apply("ReshuffleKeys", reshuffle)
+                .apply(
+                    "MergeTransformWrite",
+                    ParDo.of(new MergeAndWriteBuckets<>(
+                      sources,
+                      sourceSpec.keyCoder,
+                      tempFileAssignment,
+                      fileOperations,
+                      bucketMetadata,
+                      transformFn)
+                  )
+                ).setCoder(KvCoder.of(BucketShardIdCoder.of(), ResourceIdCoder.of()))
+        ).apply(
+            "FinalizeTempFiles",
+            new SortedBucketSink.FinalizeTempFiles<>(
+                filenamePolicy.forDestination(), bucketMetadata, fileOperations)
+        );
+  }
+
+  @FunctionalInterface
+  public interface TransformFn<KeyT, ValueT> extends Serializable {
+    void writeTransform(KV<KeyT, CoGbkResult> keyGroup, OutputCollector<ValueT> outputConsumer);
+  }
+
+  public static class OutputCollector<ValueT> implements Consumer<ValueT>, Serializable {
+    private final Writer<ValueT> writer;
+
+    public OutputCollector(Writer<ValueT> writer) {
+      this.writer = writer;
+    }
+
+    public void onComplete() {
+      try {
+        writer.close();
+      } catch (IOException e) {
+        throw new RuntimeException("Closing writer failed: ", e);
+      }
+    }
+
+    @Override
+    public void accept(ValueT t) {
+      try {
+        writer.write(t);
```

Comment:
Is it weird to add a custom `Counter` here, or are they only supposed to be added in user code? It would be nice to have some instrumentation for users to see how many elements have been written -- as is Dataflow only reports the # of buckets, there's no visibility on actual data size. I couldn't find any examples in Beam of adding Counters to standard library `DoFn`s tho.

Reply:
Interesting point. This is a special case since data was read, transformed and written in the same transform, and doesn't go through any DF estimator right? Guess we can make an exception here. But the name needs to be unique?

Reply:
yeah, there really is no helpful information in the Dataflow UI, heh. I think the name should be okay as long as it's in the `com.spotify.scio.smb` namespace? Maybe `com.spotify.scio.smb.MergedElementsWritten` ? It could be nice to add one for number of elements/key groups read, too.

Reply:
Is it possible to have multiple SMB sources per workflow though? They'll need some suffix to uniquely identify each?

Reply:
You mean multiple `SortedBucketSource`  or `SortedBucketTransform`s per workflow? Hm we could suffix with PTransform name? And maybe TupleTag of the BucketedInput if we want to measure elements read at that granularity.

How about I fix up the bucket fanout thing, merge this PR, then follow up with the API bindings/`Counter` instrumentation in separate PRs?

Reply:
Yeah sounds good!